### PR TITLE
EPMRPP-33481 Widget modals. Add shared widget modal

### PR DESCRIPTION
**Pull request by AmsterGet**
Task - https://jiraeu.epam.com/browse/EPMRPP-33481
**Features:**
* Add shared widget modal
* Rewrite the mechanism for adding widgets (recalculating widget positions on the layout).
* Widget preview component

### CODE REVIEW

File: app/src/pages/inside/dashboardItemPage/modals/addSharedWidgetModal/sharedWidgetsListSection/widgetsList/widgetsList.jsx [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    </ScrollWrapper>
+  </div>
+);
+
+WidgetsList.propTypes = {
+  activeId: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
+  widgets: PropTypes.array.isRequired,
+  loading: PropTypes.bool,
+  onChange: PropTypes.func,
+  onLazyLoad: PropTypes.oneOfType([PropTypes.func, PropTypes.bool]),
```
Comment (DiscoElevator) [inline app/src/pages/inside/dashboardItemPage/modals/addSharedWidgetModal/sharedWidgetsListSection/widgetsList/widgetsList.jsx]: func or boolean? Looks strange

File: app/src/pages/inside/dashboardItemPage/modals/common/widgetControls/controls/customColumnsControl/customColumnsControl.jsx [hunk 2]
```diff
@@ -5,7 +5,7 @@ import classNames from 'classnames/bind';
 import { ModalField } from 'components/main/modal';
 import { FIELD_LABEL_WIDTH } from '../constants';
 import styles from './customColumnsControl.scss';
-import { CustomColumnItem } from './customColumnItem';
+import { CustomColumnItem } from './customColumnItem/index';
```
Comment (DiscoElevator) [inline app/src/pages/inside/dashboardItemPage/modals/common/widgetControls/controls/customColumnsControl/customColumnsControl.jsx]: /index

File: app/src/pages/inside/dashboardItemPage/modals/common/widgetControls/controls/filtersControl/common/addEditFilter/addEditFilter.jsx [hunk 3]
```diff
@@ -7,7 +7,7 @@ import { COMMON_LOCALE_KEYS } from 'common/constants/localization';
 import { BigButton } from 'components/buttons/bigButton';
 import { FilterEntitiesContainer } from 'components/filterEntities/containers';
 import { EntitiesGroup } from 'components/filterEntities/entitiesGroup';
-import { FiltersSorting } from '../../filtersSorting';
+import { FiltersSorting } from '../../filtersSorting/index';
```
Comment (DiscoElevator) [inline app/src/pages/inside/dashboardItemPage/modals/common/widgetControls/controls/filtersControl/common/addEditFilter/addEditFilter.jsx]: /index

File: app/src/pages/inside/dashboardItemPage/modals/common/widgetControls/controls/filtersControl/filterAdd/filterAdd.jsx [hunk 4]
```diff
@@ -10,7 +10,7 @@ import { FieldErrorHint } from 'components/fields/fieldErrorHint';
 import { Input } from 'components/inputs/input';
 import { messages } from '../common/messages';
 import { FILTER_ADD_FORM, FILTER_NAME_KEY } from '../common/constants';
-import { AddEditFilter } from '../common/addEditFilter';
+import { AddEditFilter } from '../common/addEditFilter/index';
```
Comment (DiscoElevator) [inline app/src/pages/inside/dashboardItemPage/modals/common/widgetControls/controls/filtersControl/filterAdd/filterAdd.jsx]: /index

File: app/src/pages/inside/dashboardItemPage/modals/common/widgetControls/controls/filtersControl/filterEdit/filterEdit.jsx [hunk 5]
```diff
@@ -3,7 +3,7 @@ import PropTypes from 'prop-types';
 import classNames from 'classnames/bind';
 import { InputRadio } from 'components/inputs/inputRadio';
 import { messages } from '../common/messages';
-import { AddEditFilter } from '../common/addEditFilter';
+import { AddEditFilter } from '../common/addEditFilter/index';
```
Comment (DiscoElevator) [inline app/src/pages/inside/dashboardItemPage/modals/common/widgetControls/controls/filtersControl/filterEdit/filterEdit.jsx]: /index

File: app/src/pages/inside/dashboardItemPage/modals/common/widgetControls/controls/filtersControl/filtersControl.jsx [hunk 6]
```diff
@@ -14,13 +14,14 @@ import {
...
-import { WIDGET_WIZARD_FORM } from '../../../widgetWizardContent/wizardControlsSection/constants';
+import { PAGE_KEY, SIZE_KEY } from 'controllers/pagination';
+import { WIDGET_WIZARD_FORM } from '../../../../widgetWizardModal/constants';
 import { FilterNotFound } from './filterNotFound/filterNotFound';
-import { FiltersActionPanel } from './filtersActionPanel';
-import { ActiveFilter } from './activeFilter';
-import { FiltersList } from './filtersList';
-import { FilterEdit } from './filterEdit';
-import { FilterAdd } from './filterAdd';
+import { FiltersActionPanel } from './filtersActionPanel/index';
```
Comment (DiscoElevator) [inline app/src/pages/inside/dashboardItemPage/modals/common/widgetControls/controls/filtersControl/filtersControl.jsx]: /index

File: app/src/pages/inside/dashboardItemPage/modals/common/widgetControls/controls/filtersControl/filtersControl.jsx [hunk 7]
```diff
@@ -14,13 +14,14 @@ import {
   loadingSelector,
   filtersPaginationSelector,
 } from 'controllers/filter';
-import { WIDGET_WIZARD_FORM } from '../../../widgetWizardContent/wizardControlsSection/constants';
+import { PAGE_KEY, SIZE_KEY } from 'controllers/pagination';
+import { WIDGET_WIZARD_FORM } from '../../../../widgetWizardModal/constants';
```
Comment (DiscoElevator) [inline app/src/pages/inside/dashboardItemPage/modals/common/widgetControls/controls/filtersControl/filtersControl.jsx]: I think absolute path will be better here

File: app/src/pages/inside/dashboardItemPage/modals/common/widgetControls/controls/filtersControl/filtersList/filtersList.jsx [hunk 8]
```diff
@@ -4,7 +4,7 @@ import classNames from 'classnames/bind';
 import { ScrollWrapper } from 'components/main/scrollWrapper';
 import { SpinningPreloader } from 'components/preloaders/spinningPreloader/spinningPreloader';
 import styles from './filtersList.scss';
-import { FiltersItem } from '../filtersItem';
+import { FiltersItem } from '../filtersItem/index';
```
Comment (DiscoElevator) [inline app/src/pages/inside/dashboardItemPage/modals/common/widgetControls/controls/filtersControl/filtersList/filtersList.jsx]: /index

File: app/src/pages/inside/dashboardItemPage/modals/common/widgetControls/controls/index.js [hunk 9]
```diff
@@ -1,7 +1,7 @@
 export { CheckboxControl } from './checkboxControl';
 export { DropdownControl } from './dropdownControl';
-export { FiltersControl } from './filtersControl';
+export { FiltersControl } from './filtersControl/index';
```
Comment (DiscoElevator) [inline app/src/pages/inside/dashboardItemPage/modals/common/widgetControls/controls/index.js]: /index

File: app/src/controllers/dashboard/selectors.js [hunk 10]
```diff
@@ -1,4 +1,18 @@
+import { createSelector } from 'reselect';
+import { activeDashboardIdSelector } from 'controllers/pages';
+
+const domainSelector = (state) => state.dashboard || {};
+
+export const dashboardItemsUnsortedSelector = (state) => domainSelector(state).dashboardItems;
+
 export const dashboardItemsSelector = (state) =>
-  state.dashboard.dashboardItems.sort((a, b) => a.name.localeCompare(b.name));
+  dashboardItemsUnsortedSelector(state).sort((a, b) => a.name.localeCompare(b.name));
```
[the remark this answers is not in the document]
Comment (AmsterGet) [reply app/src/controllers/dashboard/selectors.js]: Fixed

File: app/src/pages/inside/dashboardItemPage/modals/addSharedWidgetModal/addSharedWidgetModal.jsx [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+
+  onAdd = (closeModal) => {
+    const {
+      data: { onConfirm },
+    } = this.props;
+    this.props.showScreenLockAction();
+    const widget = {
+      widgetId: this.state.selectedWidget.id,
+      widgetPosition: { positionX: 0, positionY: 0 },
+      widgetSize: { width: 12, height: 7 },
```
[the remark this answers is not in the document]
Comment (AmsterGet) [reply app/src/pages/inside/dashboardItemPage/modals/addSharedWidgetModal/addSharedWidgetModal.jsx]: Fixed

File: app/src/pages/inside/dashboardItemPage/modals/addSharedWidgetModal/sharedWidgetsListSection/sharedWidgetsListSection.jsx [hunk 12]
```diff
@@ -0,0 +1,155 @@
+import React, { Component } from 'react';
+import PropTypes from 'prop-types';
+import { defineMessages, injectIntl, intlShape } from 'react-intl';
+import classNames from 'classnames/bind';
+import { URLS } from 'common/urls';
+import { fetch } from 'common/utils';
+import { PAGE_KEY, SIZE_KEY } from 'controllers/pagination';
+import { COMMON_LOCALE_KEYS } from 'common/constants/localization';
+import { WidgetsList } from './widgetsList/index';
```
[the remark this answers is not in the document]
Comment (AmsterGet) [reply app/src/pages/inside/dashboardItemPage/modals/addSharedWidgetModal/sharedWidgetsListSection/sharedWidgetsListSection.jsx]: Fixed

File: app/src/pages/inside/dashboardItemPage/modals/addSharedWidgetModal/sharedWidgetsListSection/sharedWidgetsListSection.jsx [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+import { SharedWidgetsSearch } from './sharedWidgetsSearch/index';
```
[the remark this answers is not in the document]
Comment (AmsterGet) [reply app/src/pages/inside/dashboardItemPage/modals/addSharedWidgetModal/sharedWidgetsListSection/sharedWidgetsListSection.jsx]: Fixed

File: app/src/pages/inside/dashboardItemPage/modals/addSharedWidgetModal/sharedWidgetsListSection/sharedWidgetsListSection.jsx [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+  static defaultProps = {
+    intl: {},
+    selectedWidget: null,
+    onSelectWidget: () => {},
+  };
+
+  constructor(props) {
+    super(props);
+
+    this.state = {
```
[the remark this answers is not in the document]
Comment (AmsterGet) [reply app/src/pages/inside/dashboardItemPage/modals/addSharedWidgetModal/sharedWidgetsListSection/sharedWidgetsListSection.jsx]: Fixed

File: app/src/pages/inside/dashboardItemPage/modals/addSharedWidgetModal/sharedWidgetsListSection/widgetsList/widgetsList.jsx (hunk 1 above)
Comment (AmsterGet) [reply app/src/pages/inside/dashboardItemPage/modals/addSharedWidgetModal/sharedWidgetsListSection/widgetsList/widgetsList.jsx]: Fixed

File: app/src/pages/inside/dashboardItemPage/modals/common/widgetControls/controls/customColumnsControl/customColumnsControl.jsx (hunk 2 above)
Comment (AmsterGet) [reply app/src/pages/inside/dashboardItemPage/modals/common/widgetControls/controls/customColumnsControl/customColumnsControl.jsx]: Fixed

File: app/src/pages/inside/dashboardItemPage/modals/common/widgetControls/controls/filtersControl/common/addEditFilter/addEditFilter.jsx (hunk 3 above)
Comment (AmsterGet) [reply app/src/pages/inside/dashboardItemPage/modals/common/widgetControls/controls/filtersControl/common/addEditFilter/addEditFilter.jsx]: Fixed

File: app/src/pages/inside/dashboardItemPage/modals/common/widgetControls/controls/filtersControl/filterAdd/filterAdd.jsx (hunk 4 above)
Comment (AmsterGet) [reply app/src/pages/inside/dashboardItemPage/modals/common/widgetControls/controls/filtersControl/filterAdd/filterAdd.jsx]: Fixed

File: app/src/pages/inside/dashboardItemPage/modals/common/widgetControls/controls/filtersControl/filterEdit/filterEdit.jsx (hunk 5 above)
Comment (AmsterGet) [reply app/src/pages/inside/dashboardItemPage/modals/common/widgetControls/controls/filtersControl/filterEdit/filterEdit.jsx]: Fixed

File: app/src/pages/inside/dashboardItemPage/modals/common/widgetControls/controls/filtersControl/filtersControl.jsx (hunk 6 above)
Comment (AmsterGet) [reply app/src/pages/inside/dashboardItemPage/modals/common/widgetControls/controls/filtersControl/filtersControl.jsx]: Fixed

File: app/src/pages/inside/dashboardItemPage/modals/common/widgetControls/controls/filtersControl/filtersList/filtersList.jsx (hunk 8 above)
Comment (AmsterGet) [reply app/src/pages/inside/dashboardItemPage/modals/common/widgetControls/controls/filtersControl/filtersList/filtersList.jsx]: Fixed

File: app/src/pages/inside/dashboardItemPage/modals/common/widgetControls/controls/index.js (hunk 9 above)
Comment (AmsterGet) [reply app/src/pages/inside/dashboardItemPage/modals/common/widgetControls/controls/index.js]: Fixed

File: app/src/pages/inside/dashboardItemPage/modals/common/widgetControls/controls/filtersControl/filtersControl.jsx (hunk 7 above)
Comment (AmsterGet) [reply app/src/pages/inside/dashboardItemPage/modals/common/widgetControls/controls/filtersControl/filtersControl.jsx]: Fixed

**Pull request merged**